stm32-modules/heater-shaker/firmware/host_comms_task/freertos_comms_task.cpp

```diff
@@ -308,3 +332,52 @@ static auto CDC_Receive(uint8_t *Buf, uint32_t *Len) -> int8_t {
     USBD_CDC_ReceivePacket(&_local_task.usb_handle);
     return USBD_OK;
 }
+
+void HAL_UART_RxCpltCallback(UART_HandleTypeDef *UartHandle)
+{
+    using namespace host_comms_control_task;
+    uint8_t *Buf = reinterpret_cast<uint8_t *>(_local_task.uart_rx_buf.committed()->data());
+    uint16_t Len = (uint16_t)(_local_task.uart_rx_buf.committed()->size());
+    uint8_t *buffer_count = std::find_if_not(Buf + Len, Buf, [](auto ch) { return ch == '0'; });
+    ssize_t remaining_buffer_count = reinterpret_cast<ssize_t>(Buf + Len - buffer_count);
+    // NOLINTNEXTLINE(cppcoreguidelines-pro-bounds-pointer-arithmetic)
+    if ((std::find_if(Buf, Buf + Len,
+                      [](auto ch) { return ch == '\n' || ch == '\r'; }) !=
+         (Buf +  // NOLINT(cppcoreguidelines-pro-bounds-pointer-arithmetic)
+          Len)) ||
+        remaining_buffer_count <
+            static_cast<ssize_t>(CDC_DATA_HS_MAX_PACKET_SIZE)) {
+        // there was a newline in this message, can pass on
+        auto message =
+            messages::HostCommsMessage(messages::IncomingMessageFromHost{
+                // NOLINTNEXTLINE(cppcoreguidelines-pro-type-reinterpret-cast)
+                .buffer = reinterpret_cast<const char *>(
+                    _local_task.rx_buf.committed()->data()),
+                // NOLINTNEXTLINE(cppcoreguidelines-pro-type-reinterpret-cast)
+                .limit = reinterpret_cast<const char *>(
+                    // NOLINTNEXTLINE(cppcoreguidelines-pro-bounds-pointer-arithmetic)
+                    buffer_count)});
+        static_cast<void>(
+            _top_task.get_message_queue().try_send_from_isr(message));
+        _local_task.rx_buf.swap();
+        _local_task.committed_rx_buf_ptr =
+            _local_task.rx_buf.committed()->data();
+    } else {
+        // NOLINTNEXTLINE(cppcoreguidelines-pro-bounds-pointer-arithmetic)
+        _local_task.committed_rx_buf_ptr += (buffer_count - Buf);
+    }
+    
+    HAL_UART_Receive_IT(
+        &_local_task.uart_handle,
+        // NOLINTNEXTLINE(cppcoreguidelines-pro-type-reinterpret-cast)
+        reinterpret_cast<uint8_t *>(
+            _local_task.uart_rx_buf.committed()->data()),
```

Comment:
this has to be `_local_task.committed_rx_buf_ptr`. the logic in the block above does the following:
- See if there's a newline here indicating that the command is completed, or the buffer is out of space
- If so, send it to the task
- If not, advance the buffer input cursor past what was received so we append the next data to it

I think this isn't actually used in the USB side because USB is packetized enough that you really don't get incomplete commands often, but it's important here - we're gonna be getting a lot of interrupts that have like one character in the input buffer.

So here we want to use `_local_task.committed_rx_buf_ptr` for the pointer and probably add something to keep track of the remaining size.